java/ql/src/semmle/code/java/JDK.qll

```diff
@@ -291,6 +296,46 @@ class MethodMathMax extends Method {
   }
 }
 
+/**
+ * Any method named `compile` on class `java.util.regex.Pattern`.
+ */
+class MethodPatternCompile extends Method {
+  MethodPatternCompile() {
+    this.hasName("compile") and
+    this.getDeclaringType() instanceof TypePattern
+  }
+}
+
+/**
+ * Any method named `matches` on class `java.util.regex.Pattern`.
+ */
+class MethodPatternMatches extends Method {
+  MethodPatternMatches() {
+    this.hasName("matches") and
+    this.getDeclaringType() instanceof TypePattern
+  }
+}
+
+/**
+ * Any method named `matcher` on class `java.util.regex.Pattern`.
+ */
+class MethodPatternMatcher extends Method {
+  MethodPatternMatcher() {
+    this.hasName("matcher") and
+    this.getDeclaringType() instanceof TypePattern
+  }
+}
+
+/**
+ * Any method named `matches` on class `java.lang.String`.
+ */
+class MethodStringMatches extends Method {
+  MethodStringMatches() {
+    this.hasName("matches") and
+    this.getDeclaringType() instanceof TypeString
+  }
+}
+
 // --- Standard fields ---
```

Comment:
Would be nice to cover common utility libraries such as Apache Commons. e.g.
- https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/RegExUtils.html
- https://commons.apache.org/proper/commons-validator/apidocs/org/apache/commons/validator/routines/RegexValidator.html
- https://commons.apache.org/proper/commons-digester/commons-digester-3.0/apidocs/org/apache/commons/digester3/SimpleRegexMatcher.html
- https://commons.apache.org/proper/commons-lang/apidocs/org/apache/commons/lang3/Validate.html#matchesPattern-java.lang.CharSequence-java.lang.String-
- ...

Also, any Spring related classes if any